[Dataset hub] more than storage: streaming, editing, connectors (#2052)
huggingface/hub-docs: docs/hub/datasets-libraries.md

Therefore, they can run end-to-end streaming pipelines: streaming from a source and writing to Hugging Face progressively, often overlapping the download, upload, and processing steps.

For more details on how to do streaming, check out the documentation of a library that support streaming (see table above) or the [streaming datasets](./datasets-streaming) documentation if you want to stream datasets from Hugging Face by yourself.

### Optimized Parquet files

Parquet files on Hugging Face are optimized to improve storage efficiency, accelerate downloads and uploads, and enable efficient dataset streaming and editing:

* [Parquet Content Defined Chunking](https://huggingface.co/blog/parquet-cdc) optimizes Parquet for [Xet](https://huggingface.co/docs/hub/en/xet/index), Hugging Face's storage backend. It accelerates uploads and downloads thanks to chunk-based deduplication and allows efficient file editing
* Page index accelerates filters when streaming and enables efficient random access, e.g. in the [Dataset Viewer](https://huggingface.co/docs/dataset-viewer)

Some libraries require extra argument to write optimized Parquet files like `Pandas` and `PyArrow`:

> * `content_defined_chunking=True` to enable Parquet Content Defined Chunking, for [deduplication](https://huggingface.co/blog/parquet-cdc) and [editing](./datasets-editing)

Reviewer: use_content_defined_chunking=True

Thanks to the [huggingface_hub Python library](/docs/huggingface_hub), it's easy to enable sharing your datasets on the Hub.
We're happy to welcome to the Hub a set of Open Source libraries that are pushing Machine Learning forward.

## Libraries table

The table below summarizes the supported libraries and their level of integration.

| Library                             | Description                                                                                                                    | Download from Hub | Push to Hub |
| ----------------------------------- | ------------------------------------------------------------------------------------------------------------------------------ | ----------------- | ----------- |
| [Argilla](./datasets-argilla)       | Collaboration tool for AI engineers and domain experts that value high quality data.                                           | ✅                | ✅          |

> | [Daft](./datasets-daft)             | Data engine for large scale, multimodal data processing with a Python-native interface.                                        | ✅ +s             | ✅ +s +p    |

Author: I liked the current approach because there is more space and proportionally more ✅ than ❌ than with more columns, but with columns dedicated to streaming I feel like we put more emphasis on streaming and optimized parquet features which is great. I'll do the change and check how it looks

> | [Daft](./datasets-daft)             | Data engine for large scale, multimodal data processing with a Python-native interface.                                        | ✅ +s             | ✅ +s +p    |

Author: after:

<img width="1728" height="957" alt="image" src="https://github.com/user-attachments/assets/2ac15138-4465-4236-ac05-b2a28edcdc5f" />

| [Pandas](./datasets-pandas)         | Python data analysis toolkit.                                                                                                  | ✅                | ✅ +p*      |
| [Polars](./datasets-polars)         | A DataFrame library on top of an OLAP query engine.                                                                            | ✅ +s             | ✅          |
| [PyArrow](./datasets-pyarrow)       | Apache Arrow is a columnar format and a toolbox for fast data interchange and in-memory analytics.                             | ✅ +s             | ✅ +p*      |
| [Spark](./datasets-spark)           | Real-time, large-scale data processing tool in a distributed environment.                                                      | ✅ +s             | ✅ +s +p    |
| [WebDataset](./datasets-webdataset) | Library to write I/O pipelines for large datasets.                                                                             | ✅ +s             | ❌          |

_+s: Supports Streaming_
_+p: Writes optimized Parquet files_

> _+p*: Requires passing extra arguments to write optimized Parquet files_

Reviewer 2: nice!

### Streaming

Dataset streaming allows to iterate on a dataset on Hugging Face progressively without having to download it completely.

> It saves disk space and download time.

Reviewer 3: curious to know, how does it save download time?

Author: Maybe it's a bit unclear. You don't need to spend time downloading data in advance since it streams in parallel to the cpu / gpu workload. Let me add general clarifications here:

```suggestion
It saves disk space since the data is never materialized on disk. It saves memory since only a small portion of the dataset is used at a time. And it saves time, since there is no need to download data in advance prior to the actual CPU or GPU workload.
```

> It saves disk space since the data is never materialized on disk. It saves memory since only a small portion of the dataset is used at a time. And it saves time, since there is no need to download data in advance prior to the actual CPU or GPU workload.

Reviewer 4: ```suggestion
It saves local disk space because the entire data is never on disk. It saves memory since only a small portion of the dataset is used at a time. And it saves time, since there is no need to download data before the CPU or GPU workload.
```

In addition to streaming from Hugging Face, many libraries also support streaming when writing back to Hugging Face.

> Therefore they can run end-to-end streaming pipelines: streaming from a source and writing to Hugging Face progressively, often overlapping the downloads, uploads and processing steps.

Reviewer 2: ```suggestion
Therefore, they can run end-to-end streaming pipelines: streaming from a source and writing to Hugging Face progressively, often overlapping the download, upload, and processing steps.
```

For more details on how to do streaming, check out the documentation of a library that support streaming (see table above) or the [streaming datasets](./datasets-streaming) documentation if you want to stream datasets from Hugging Face by yourself.

### Optimized Parquet files

> Parquet files on Hugging Face are optimized to improve storage efficiency, accelerate downloads and uploads, and enable efficient dataset streaming and editing:

Reviewer 5: [internal issue for discussion](https://github.com/huggingface-internal/moon-landing/issues/15793)

> * [Parquet Content Defined Chunking](https://huggingface.co/blog/parquet-cdc) optimizes Parquet for [Xet](https://huggingface.co/docs/hub/en/xet/index), Hugging Face's storage based on Git. It accelereates uploads and downloads thanks to deduplication and allows efficient file editing

Reviewer 3: ```suggestion
* [Parquet Content Defined Chunking](https://huggingface.co/blog/parquet-cdc) optimizes Parquet for [Xet](https://huggingface.co/docs/hub/en/xet/index), Hugging Face's storage backend. It accelerates uploads and downloads thanks to chunk-based deduplication and allows efficient file editing
```
